### PR TITLE
Fix Storage invalid firmware version read #960

### LINKED ISSUE #960: Nitrokey Storage read firmware version is 0.0
Version: v1.2.3
Same details as in: https://github.com/hughsie/fwupd/issues/949

#### Current behaviour
Nitrokey plugin for reading the firmware version from the device seems to return always `0.0`.

#### Expected behaviour
Nitrokey plugin for reading the firmware version should return the correct value.

### PR DESCRIPTION
**Pull request by szszszsz**
This patch corrects the Nitrokey Storage firmware version getter plugin. 
Commit message:
>  Correct Nitrokey Storage invalid firmware version read
Currently used structures were based on early libnitrokey definitions
(which were broken for some time). Corrected their sizes and elements,
and added tests.

> Tested live on Fedora 29, with Nitrokey Storage v0.53.
Bundled tests pass on my build.

Fixes #960

Type of pull request:
- [ ] New plugin 
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/nitrokey/fu-nitrokey-device.c [hunk 1]
```diff
@@ -46,10 +46,10 @@ typedef struct __attribute__((packed)) {
...
 	guint8		ReadWriteFlagUncryptedVolume;
 	guint8		ReadWriteFlagCryptedVolume;
-	guint8		VersionReserved1;
-	guint8		VersionMinor;
-	guint8		VersionReserved2;
-	guint8		VersionMajor;
+	guint8      VersionMajor;
+	guint8      VersionMinor;
+	guint8      VersionReservedByte;
+	guint8      VersionBuildIteration;
```
Comment (superm1) [inline plugins/nitrokey/fu-nitrokey-device.c]: whitespace doesn't look lined up

File: plugins/nitrokey/fu-self-test.c [hunk 2]
```diff
@@ -10,6 +10,29 @@
...
+    0x03,0x03,0xc7,0x37,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x14,0x2d,0x76,
+    0xa2
+    };
+
+    _dump_to_console ("payload", buf, sizeof(buf));
+    NitrokeyGetDeviceStatusPayload payload;
+    memcpy (&payload, buf, sizeof(buf));
+    g_assert_cmpint (payload.VersionMajor, ==, 0);
+    g_assert_cmpint (payload.VersionMinor, !=, 0);
+    g_assert_cmpint (payload.VersionMinor, ==, 53);
```
Comment (superm1) [inline plugins/nitrokey/fu-self-test.c]: I think you only need to have one check on `VersionMinor`  If it matches 53 you know it's not 0.

Comment (szszszsz) [reply plugins/nitrokey/fu-self-test.c]: I wanted to make sure about this ;-) You are right of course, the `!= 0` condition will be removed.

File: plugins/nitrokey/fu-nitrokey-device.c (hunk 1 above)
Comment (szszszsz) [reply plugins/nitrokey/fu-nitrokey-device.c]: Confirmed. Will be corrected.

File: plugins/nitrokey/fu-self-test.c (hunk 2 above)
Comment (hughsie) [reply plugins/nitrokey/fu-self-test.c]: Cool, if you push the branch again it looks good to me with these two fixes. Thanks for tracking it down!

**Comment by szszszsz** [comment]
I wanted to test my patch live on my dev Fedora 29 system. I have installed the package, that Docker built for me (as done in CI). Version reported is correct (1.2.4). Reported device firmware version is still `0.0` though - is this value cached somewhere?
I am using a different version of code - will upload in a minute.

**Comment by szszszsz** [comment]
Updated code. Commits prefixed with `REMOVE` will be removed entirely later. These contain changes for the local build environment, which are not in connection with the issue being solved.

**Comment by szszszsz** [comment]
Ready for review.
I might have messed the formatting a bit in the plugin's tests. I could not adjust my editor to match the old one for some reason.

**Comment by hughsie** [comment]
With the old code, my version was `0.48` -- with the new code (struct fixes) the version is the more plausible `0.53`. I've rebased this, fixed a few whitespace errors and pushed it to master. Many thanks for the debugging.

**Comment by szszszsz** [comment]
Great! Thank you! Do you know perhaps the date of the next `fwupd` release?

**Comment by hughsie** [comment]
We normally release one per month, the last release being on the 4th, so probably in the next week or so.

**Pull request closed without merging**